File: src/main/java/ru/fess38/finance/dao/UserDaoImpl.java
package ru.fess38.finance.dao;

import org.hibernate.SessionFactory;
import org.hibernate.criterion.DetachedCriteria;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;
import ru.fess38.finance.DatabaseEventListener;
import ru.fess38.finance.model.ModifiableUser;
import ru.fess38.finance.model.User;

import java.util.List;
import java.util.stream.Collectors;

@Repository
@Transactional
public class UserDaoImpl implements UserDao {
  private SessionFactory sessionFactory;
  private DatabaseEventListener databaseEventListener;

  @Override
  public User save(User user) {
    return update(user);
  }

  @Override
  public User get(long id) {
    return sessionFactory.getCurrentSession().get(ModifiableUser.class, id).toImmutable();
  }

  @Override
  public User update(User user) {
    ModifiableUser modifiableUser = (ModifiableUser) sessionFactory.getCurrentSession()
        .merge(user.toModifiable());
    databaseEventListener.setChangeTrue();
    return modifiableUser.toImmutable();
  }

  @Override
  public User delete(User user) {
    return update(user.withIsDeleted(true));
  }

  @Override
  public List<User> find(DetachedCriteria detachedCriteria) {
    return find(detachedCriteria, false);
  }

  @Override
  public List<User> findDeleted(DetachedCriteria detachedCriteria) {
    return find(detachedCriteria, true);
  }

  @SuppressWarnings("unchecked")
  private List<User> find(DetachedCriteria detachedCriteria, boolean isDeleted) {
    detachedCriteria = isDeleted ? DaoHelper.deleted(detachedCriteria) :
        DaoHelper.notDeleted(detachedCriteria);
    return (List<User>) detachedCriteria
        .getExecutableCriteria(sessionFactory.getCurrentSession())
        .list()
        .stream()
        .map(x -> ((ModifiableUser) x).toImmutable())
        .collect(Collectors.toList());
  }

  @Override
  public DetachedCriteria detachedCriteria() {
    return DetachedCriteria.forClass(ModifiableUser.class);
  }

  @Autowired
  public void setSessionFactory(SessionFactory sessionFactory) {
    this.sessionFactory = sessionFactory;
  }

  @Autowired
  public void setDatabaseEventListener(DatabaseEventListener databaseEventListener) {
    this.databaseEventListener = databaseEventListener;
  }
}

File: kotlin/src/main/kotlin/ru/fess38/finance/security/UserDaoImpl.kt
package ru.fess38.finance.security

import org.hibernate.SessionFactory
import org.hibernate.criterion.DetachedCriteria
import org.hibernate.criterion.Restrictions
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional
import ru.fess38.finance.core.Model.RefreshToken.AuthType

@Repository
@Transactional
class UserDaoImpl: UserDao {
  @Autowired
  lateinit var sessionFactory: SessionFactory

  override fun getAll(): List<User> {
    val criteria = DetachedCriteria.forClass(User::class.java, "user")
    val session = sessionFactory.openSession()
    return criteria.getExecutableCriteria(session).list()
        .map {it as User}
        .distinct()
        .apply {session.close()}
  }

  override fun save(user: User): User {
    sessionFactory.openSession().apply {save(user)}.apply {flush()}.apply {close()}
    return user
  }

  override fun update(user: User) {
    sessionFactory.openSession().apply {update(user)}.apply {flush()}.apply {close()}
  }

  override fun find(token: String): User? {
    val criteria = DetachedCriteria.forClass(User::class.java, "user")
        .createAlias("user.sessions", "s")
        .add(Restrictions.eq("s.token", token))
        .add(Restrictions.gt("s.expired", System.currentTimeMillis()))
    val session = sessionFactory.openSession()
    return criteria.getExecutableCriteria(session).list()
        .map {it as User}
        .firstOrNull()
        .apply {session.close()}
  }

  override fun find(outerId: String, authType: AuthType): User? {
    val criteria = DetachedCriteria.forClass(User::class.java)
        .add(Restrictions.eq("outerId", outerId))
        .add(Restrictions.eq("authType", authType))
    val session = sessionFactory.openSession()
    return criteria.getExecutableCriteria(session).list()
        .map {it as User}
        .firstOrNull()
        .apply {session.close()}
  }
}
